cirq-core/cirq/transformers/routing/route_circuit_cqc.py

```diff
@@ -0,0 +1,327 @@
+                if protocols.num_qubits(op) == 2:
+                    return op
+                single_qubit_ops_preserve[moment_index].append(op)
+                return []
+
+            reduced_circuit_preserve = transformer_primitives.map_operations(
+                circuit, map_func=map_func_preserving
+            )
+            return (
+                [list(moment.operations) for moment in reduced_circuit_preserve],
+                single_qubit_ops_preserve,
+            )
+
+        reduced_circuit = circuits.Circuit()
+
+        def map_func_not_preserving(op: 'cirq.Operation', _: int):
+            timestep_index = reduced_circuit.earliest_available_moment(op)
+            if protocols.num_qubits(op) == 2:
+                reduced_circuit.append(op)
+            return op.with_tags(timestep_index)
+
+        circuit_with_tags = transformer_primitives.map_operations(
+            circuit, map_func=map_func_not_preserving
+        )
+        single_qubit_operations: List[List['cirq.Operation']] = [
+            [] for i in range(len(reduced_circuit) + 1)
+        ]
+        for op in circuit_with_tags.all_operations():
+            if protocols.num_qubits(op) == 1 and isinstance(op.tags[-1], int):
+                timestep_index = op.tags[-1]
+                single_qubit_operations[timestep_index].append(op)
+        return [list(moment.operations) for moment in reduced_circuit], single_qubit_operations
+
+    def _route(
+        self,
+        timesteps: List[List['cirq.Operation']],
+        single_qubit_ops: List[List['cirq.Operation']],
+        max_search_radius: int,
+        tag_inserted_swaps: bool = False,
+    ) -> List[List['cirq.Operation']]:
+        """Main routing procedure that creates the routed circuit, inserts the user's gates in it,
+        and inserts the necessary swaps.
+
+        Args:
+          timesteps: the circuit's timesteps as defined by the paper
+          max_search_radius: the maximum number of times the cost function can be iterated for
+            convergence.
+
+        Returns:
+          a list of lists corresponding to moments of the routed circuit
+        """
+
+        def process_executable_ops(idx: int):
+            unexecutable_ops = []
+            for op in timesteps[idx]:
+                if self._mm.can_execute(op):
+                    routed_ops[idx].append(self._mm.mapped_op(op))
+                else:
+                    unexecutable_ops.append(op)
+            timesteps[idx] = unexecutable_ops
+
+        routed_ops: List[List['cirq.Operation']] = [[] for i in range(len(timesteps) + 1)]
+        for idx in range(len(timesteps)):
+            # add single qubit ops the current output moment.
+            routed_ops[idx].extend([self._mm.mapped_op(op) for op in single_qubit_ops[idx]])
+
+            process_executable_ops(idx)
+
+            seen: Set[Tuple[Tuple['cirq.Qid', cirq.Qid], ...]] = set()
```

Comment:
Declare a `QidPair = Tuple[cirq.Qid, cirq.Qid]`  types on top of the file and use it in defining types of objects here. The multiple recursive nesting of tuples is confusing. 

```suggestion
            seen: Set[Tuple[QidPair, ...]] = set()
```